fix(utxo): correct block header deserialization for AuxPow and KAWPOW coins

shamardy opened the pull request:
Resolves #2048

This pull request addresses a critical `UnexpectedEnd` error that caused swap failures for several UTXO-based coins, including Namecoin (NMC), Cheetahcoin (CHTA), and eCash (XEC).

## The Problem

The root cause was an overly broad conditional logic within the `BlockHeader::deserialize` function. This single function attempted to parse multiple different block header formats using `if` conditions based on the block `version` number. This led to two primary failure modes:

### 1. Incorrect `KAWPOW` Parsing
Coins like CHTA use the same `version` number as Ravencoin (`0x30000000`) for their KAWPOW implementation but do not include the extra header fields (`n_height`, `n_nonce_u64`, `mix_hash`) that RVN does. The deserializer was incorrectly attempting to read these non-existent fields for CHTA, causing a stream read to fail with `UnexpectedEnd`.

### 2. Brittle AuxPow Detection
The logic for detecting Auxiliary Proof-of-Work (AuxPow) headers relied on a hardcoded list of exact version numbers. The correct method is to check for a specific bit flag (`1 << 8`) in the version field, as the base version can change. This made the check fragile and prone to failure if a block with a different base version but an active AuxPow flag was encountered.

## The Solution

This patch refactors the deserialization logic to be more precise and robust:

### 1. Specialized KAWPOW Logic
- A new `CoinVariant::RVN` has been added in `mm2src/mm2_bitcoin/serialization/src/reader.rs`
- The deserialization logic in `mm2src/mm2_bitcoin/chain/src/block_header.rs` now explicitly checks `reader.coin_variant().is_rvn()` before attempting to parse RVN-specific header fields
- This ensures that other coins using the KAWPOW version number are not parsed incorrectly

### 2. Correct AuxPow Detection
- The check for AuxPow headers has been changed from a hardcoded `matches!` macro to the correct bitmask check: `(version & AUXPOW_VERSION_FLAG) != 0`
- This aligns with the merged mining specification and correctly identifies all AuxPow-enabled blocks

### 3. Added Debug Utility Test
- A new, ignored test, `test_scan_and_deserialize_block_headers`, has been added to `mm2src/coins/utxo/utxo_tests.rs`
- This powerful utility allows developers to quickly scan the blockchain of any UTXO coin to pinpoint the exact block height where deserialization fails
- This will significantly speed up debugging similar issues in the future

## Impact

These changes resolve the swap failures for the affected coins and make the UTXO header deserialization logic significantly more robust and maintainable.

shamardy committed:
fix(utxo): Correct block header deserialization for AuxPow and KAWPOW coins

Swaps for coins like NMC, CHTA, and XEC were failing with an
`UnexpectedEnd` error during MTP steps. This was due to
overly broad conditional logic in `BlockHeader::deserialize` which
misinterpreted block header structures based on version numbers.

This commit corrects the logic in two ways:

1.  The check for AuxPow headers is changed from a hardcoded list of
    versions to the correct bitmask check (`version & (1 << 8)`). This
    is the proper way to detect AuxPow and makes the logic robust against
    base version changes.

2.  KAWPOW header parsing is now specialized to apply only to Ravencoin.
    A `CoinVariant::RVN` is introduced, and the deserializer now checks
    for this variant before attempting to read RVN-specific fields. This
    prevents conflicts with other coins (like CHTA) that use the same
    version number but have a different header structure.

A utility test for scanning and debugging block headers has also been
added to assist with future diagnostics.

shamardy committed:
Merge remote-tracking branch 'origin/dev' into bugfix/2048-unexpected-end-error

shamardy committed:
fix clippy

mariocynicys commented on mm2src/coins/utxo/utxo_tests.rs:
```diff
@@ -5706,3 +5706,95 @@ fn test_electrum_v14_block_hash() {
+            num_to_fetch = num_to_fetch.min(remaining);
+        }
+
+        if num_to_fetch == 0 {
+            break;
+        }
```
don't think this is ever gonna trigger.
`remaining` is guaranteed to be greater than `0` and `CHUNK_SIZE` should also not be zero.

mariocynicys commented on mm2src/coins/utxo/utxo_tests.rs:
```diff
@@ -5706,3 +5706,95 @@ fn test_electrum_v14_block_hash() {
+                // If a read fails, we've found the problematic header.
+                // We panic with all the necessary context for debugging.
+                let chunk_hex_str = hex::encode(raw_chunk_bytes);
+                panic!(
+                    "\n\n!!! Deserialization failed on header index {} (block height: {}) within the chunk starting at {} !!!\n\
+                     Raw Chunk Hex: {}\n\
+                     Deserialization Error: {:?}\n\n",
```
format-wise, i think we should remove the space before `Raw Chunk ..` & `Deserialization Error..` so they don't get extra indented in the terminal.
i think we should also print the deserialization error first as the chunk_hex will probably be so big.

mariocynicys commented on mm2src/mm2_bitcoin/chain/src/block_header.rs:
```diff
@@ -242,7 +241,7 @@ impl Deserializable for BlockHeader {
             BlockHeaderNonce::H256(reader.read()?)
-        } else if (version == KAWPOW_VERSION && !reader.coin_variant().is_btc())
+        } else if (version == KAWPOW_VERSION && reader.coin_variant().is_rvn())
             || version == MTP_POW_VERSION && time >= PROG_POW_SWITCH_TIME
```
`&&` gets computed before `||` (https://doc.rust-lang.org/reference/expressions.html#r-expr.precedence)
let's put some parentheses around
`version == MTP_POW_VERSION && time >= PROG_POW_SWITCH_TIME`
to be consistent or remove the parentheses from
`(version == KAWPOW_VERSION && reader.coin_variant().is_rvn())`

mariocynicys commented on mm2src/mm2_bitcoin/chain/src/block_header.rs:
```diff
@@ -242,7 +241,7 @@ impl Deserializable for BlockHeader {
         let nonce = if is_zcash {
             BlockHeaderNonce::H256(reader.read()?)
-        } else if (version == KAWPOW_VERSION && !reader.coin_variant().is_btc())
+        } else if (version == KAWPOW_VERSION && reader.coin_variant().is_rvn())
```
this went from very broad (`!btc`) to very specific (`==rvn`). how does a coin like CHTA match this case?
i see we already have a test for CHTA so it gotta work, but wouldn't this break for some other coins?

shamardy replied:
`!reader.coin_variant().is_btc()` was added very broad wrongly in the past, it was while testing complete headers download and deserialization for BTC SPV for lightning security https://github.com/KomodoPlatform/komodo-defi-framework/pull/1452 but it turns out only rvn (that I am currently aware of) that uses `KAWPOW_VERSION` this way, and when CHTA used it, it deserialized in a wrong way as it's not BTC. About this breaking other coins, it might, but we can only rely on the test vectors / cases and if we face cases that break because of that, fix them and add them to the test cases until we have enough coverage. The alternative is to look at the codebase of each utxo coin we support and see if they use a standard header nonce with `KAWPOW_VERSION` or not, I chose not to do that 😅

shamardy committed:
apply suggestions from the @mariocynicys review.

- In `utxo_tests.rs`:
  - Remove an unreachable `break` condition from the header scanning loop.
  - Reformat the panic message in the `test_scan_and_deserialize_block_headers` utility to improve terminal readability by reordering and removing extra indentation.

- In `block_header.rs`:
  - Add parentheses to a conditional in the `BlockHeader` deserialization logic to enhance code clarity and consistency.

shamardy commented:
@mariocynicys all your review comments have been addressed :)

mariocynicys commented on mm2src/coins/utxo/utxo_tests.rs:
```diff
@@ -5706,3 +5706,91 @@ fn test_electrum_v14_block_hash() {
+                let chunk_hex_str = hex::encode(raw_chunk_bytes);
+                panic!(
+                    "\n\n!!! Deserialization failed on header index {} (block height: {}) within the chunk starting at {} !!!\n\
+                    Deserialization Error: {:?}\n\
+                    Raw Chunk Hex: {}\n\n",
```
nit: that's still indented

mariocynicys reviewed (dismissed):
LGTM!
Thanks

shamardy committed:
remove indentation one more time

shamardy committed:
fix failing `rvn_mtp` test

cipig commented:
i got an "UnexpectedEnd" error when i called recover_funds_of_swap on this swap on maker:
```
{
   "error_events" : [
      "StartFailed",
      "NegotiateFailed",
      "TakerFeeValidateFailed",
      "MakerPaymentTransactionFailed",
      "MakerPaymentDataSendFailed",
      "MakerPaymentWaitConfirmFailed",
      "TakerPaymentValidateFailed",
      "TakerPaymentWaitConfirmFailed",
      "TakerPaymentSpendFailed",
      "TakerPaymentSpendConfirmFailed",
      "MakerPaymentWaitRefundStarted",
      "MakerPaymentRefundStarted",
      "MakerPaymentRefunded",
      "MakerPaymentRefundFailed",
      "MakerPaymentRefundFinished"
   ],
   "events" : [
      {
         "event" : {
            "data" : {
               "lock_duration" : 7800,
               "maker_amount" : "0.0007",
               "maker_coin" : "PIVX",
               "maker_coin_htlc_pubkey" : "0315d9c51c657ab1be4ae9d3ab6e76a619d3bccfe830d5363fa168424c0d044732",
               "maker_coin_start_block" : 4903981,
               "maker_payment_confirmations" : 5,
               "maker_payment_lock" : 1747549945,
               "maker_payment_requires_nota" : false,
               "maker_payment_trade_fee" : {
                  "amount" : "0.001",
                  "coin" : "PIVX",
                  "paid_from_trading_vol" : false
               },
               "my_persistent_pub" : "0315d9c51c657ab1be4ae9d3ab6e76a619d3bccfe830d5363fa168424c0d044732",
               "p2p_privkey" : null,
               "secret" : "c8a43ef5478519895e8affdde0ec99e9c3a0ef6ab9ceb7324d72863c21c4b02c",
               "secret_hash" : "cd9e7193fbf9b6d75ae7b3026f394f55dd6f4b6c",
               "started_at" : 1747534345,
               "taker" : "a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025",
               "taker_amount" : "0.00081667875464",
               "taker_coin" : "KMD",
               "taker_coin_htlc_pubkey" : "0315d9c51c657ab1be4ae9d3ab6e76a619d3bccfe830d5363fa168424c0d044732",
               "taker_coin_start_block" : 4441578,
               "taker_payment_confirmations" : 3,
               "taker_payment_requires_nota" : false,
               "taker_payment_spend_trade_fee" : {
                  "amount" : "0.00001",
                  "coin" : "KMD",
                  "paid_from_trading_vol" : true
               },
               "uuid" : "be42d8c0-127c-4df5-800b-4de8318e1bf5"
            },
            "type" : "Started"
         },
         "timestamp" : 1747534345973
      },
      {
         "event" : {
            "data" : {
               "maker_coin_htlc_pubkey" : "03a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025",
               "maker_coin_swap_contract_addr" : null,
               "taker_coin_htlc_pubkey" : "03a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025",
               "taker_coin_swap_contract_addr" : null,
               "taker_payment_locktime" : 1747542145,
               "taker_pubkey" : "000000000000000000000000000000000000000000000000000000000000000000"
            },
            "type" : "Negotiated"
         },
         "timestamp" : 1747534361977
      },
      {
         "event" : {
            "data" : null,
            "type" : "MakerPaymentInstructionsReceived"
         },
         "timestamp" : 1747534363986
      },
      {
         "event" : {
            "data" : {
               "tx_hash" : "2c46c0cde1777107c58af82dc4e67ad6309589ac5226a932989412905ef996c7",
               "tx_hex" : "0400008085202f8901160660cb900ba86531faed0545967b1e61b35b8d942b1fb5ff67f15887c7af68020000006a473044022060e535195b3349aa0dcd951b5531dc44c9d6e31ac1461228d4d592585eea8e8602201d1f824c280853cdf02189c0f5e63837d9742f6d705b16b93bddc4a43e389788012103a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025ffffffff02e8030000000000001976a914ca1e04745e8ca0c60d8c5881531d51bec470743f88accc8a5a08000000001976a914ae0200524d21d348b479b30e7b93d72451fdc55688ac193a2968000000000000000000000000000000"
            },
            "type" : "TakerFeeValidated"
         },
         "timestamp" : 1747534363987
      },
      {
         "event" : {
            "data" : {
               "tx_hash" : "90e5e1199e20a5116409907756b1a6f90dcfb852fc9dced0989edfe5c5b646a2",
               "tx_hex" : "0100000001041f6dc7ec3064ff6ed5239aa979efb9866dce9693266a4321003a4f5e61c69e020000006a47304402203f9e6e763473cdd06a4a0851036f3be9cb05a6ae13e886387a62553136725ee502201edff1dccc307c848afcac493a4526c0807874120a41ad6d0feb8ba97c968e3401210315d9c51c657ab1be4ae9d3ab6e76a619d3bccfe830d5363fa168424c0d044732ffffffff03701101000000000017a914a07d020c701d6a1f4ae33a92804725a19c3dbbf1870000000000000000166a14cd9e7193fbf9b6d75ae7b3026f394f55dd6f4b6c61e06b7c110000001976a9141462c3dd3f936d595c9af55978003b27c250441f88ac1c422968"
            },
            "type" : "MakerPaymentSent"
         },
         "timestamp" : 1747534364111
      },
      {
         "event" : {
            "data" : {
               "tx_hash" : "44d8f4c0c12ca83d81a4c1a57a29371d245857e964f0d197023a2e637c4d62cf",
               "tx_hex" : "0400008085202f8901c796f95e9012949832a92652ac899530d67ae6c42df88ac5077177e1cdc0462c010000006a47304402205fc888cc95d53c649712c24d18be626e752c01fc3c5a350dc3deab101845eecf02205d3b6cbd7c1d6143d4f8c9975a28a5a1ad46be611c0b7e18f89c017227aa1f68012103a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025ffffffff03033f01000000000017a914c61e4c6c0ee6faa4ccdfff6c937a3e8d487fc736870000000000000000166a14cd9e7193fbf9b6d75ae7b3026f394f55dd6f4b6ce1475908000000001976a914ae0200524d21d348b479b30e7b93d72451fdc55688acc93b2968000000000000000000000000000000"
            },
            "type" : "TakerPaymentReceived"
         },
         "timestamp" : 1747534800903
      },
      {
         "event" : {
            "type" : "TakerPaymentWaitConfirmStarted"
         },
         "timestamp" : 1747534800903
      },
      {
         "event" : {
            "data" : {
               "error" : "maker_swap:1063] !taker_coin.wait_for_confirmations: rpc_clients:139] Tx 44d8f4c0c12ca83d81a4c1a57a29371d245857e964f0d197023a2e637c4d62cf was not found on chain after 10 tries, error: client:909] JsonRpcError { client_info: \"coin: KMD\", request: JsonRpcRequest { jsonrpc: \"2.0\", id: 8958571, method: \"blockchain.transaction.get\", params: [String(\"44d8f4c0c12ca83d81a4c1a57a29371d245857e964f0d197023a2e637c4d62cf\"), Bool(true)] }, error: Response(electrum1.cipig.net:10001, Object({\"code\": Number(2), \"message\": String(\"daemon error: DaemonError({'code': -5, 'message': 'No information available about transaction'})\")})) }"
            },
            "type" : "TakerPaymentWaitConfirmFailed"
         },
         "timestamp" : 1747537487087
      },
      {
         "event" : {
            "data" : {
               "wait_until" : 1747553645
            },
            "type" : "MakerPaymentWaitRefundStarted"
         },
         "timestamp" : 1747537487089
      },
      {
         "event" : {
            "type" : "MakerPaymentRefundStarted"
         },
         "timestamp" : 1747537487090
      },
      {
         "event" : {
            "data" : {
               "error" : "maker_swap:1315] !maker_coin.send_maker_refunds_payment: utxo_common:1958] HTLC spend fee 100000 is greater than transaction output 70000"
            },
            "type" : "MakerPaymentRefundFailed"
         },
         "timestamp" : 1747550467167
      },
      {
         "event" : {
            "type" : "Finished"
         },
         "timestamp" : 1747550467172
      }
   ],
   "gui" : "mympm",
   "maker_amount" : "0.0007",
   "maker_coin" : "PIVX",
   "maker_coin_usd_price" : null,
   "mm_version" : "2.4.0-beta_f56917b",
   "my_order_uuid" : "f98ab227-29c9-4574-a5f9-528e0ddd5b45",
   "success_events" : [
      "Started",
      "Negotiated",
      "MakerPaymentInstructionsReceived",
      "TakerFeeValidated",
      "MakerPaymentSent",
      "TakerPaymentReceived",
      "TakerPaymentWaitConfirmStarted",
      "TakerPaymentValidatedAndConfirmed",
      "TakerPaymentSpent",
      "TakerPaymentSpendConfirmStarted",
      "TakerPaymentSpendConfirmed",
      "Finished"
   ],
   "taker_amount" : "0.00081667875464",
   "taker_coin" : "KMD",
   "taker_coin_usd_price" : null,
   "type" : "Maker",
   "uuid" : "be42d8c0-127c-4df5-800b-4de8318e1bf5"
}
```
the error message is simply `"error" : "rpc:198] RPC call failed: lp_swap:1519] saved_swap:130] maker_swap:1698] utxo_standard:453] client:1117] UnexpectedEnd"` ... the log shows the same

idk if that is the same as what the PR fixes or not
anyway, coin was PIVX and it wouldn't make sense to try recover anyway since the initial error on refund was `maker_swap:1315] !maker_coin.send_maker_refunds_payment: utxo_common:1958] HTLC spend fee 100000 is greater than transaction output 70000`, so will be the same this time

shamardy commented:
> idk if that is the same as what the PR fixes or not

The `UnexpectedEnd` can happen for other coins, this PR fixed Namecoin (NMC), Cheetahcoin (CHTA), and eCash (XEC) that were specifically mentioned in the issue and maybe some others. `UnexpectedEnd` is basically problems in block headers deserializations, now I have a utility test that can uncover the reason (the bad block) quickly. I will copy this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2563#issuecomment-3151058530 to the original issue.
